### PR TITLE
datetime feature engineering added.

### PR DESCRIPTION
**Pull request by gianpDomiziani**
As discussed in the issue #76, new columns for datetime features are added, in order to improve the learning phase.

### CODE REVIEW

File: flaml/data.py [hunk 1]
```diff
@@ -156,11 +156,11 @@ def get_output_from_log(filename, time_budget):
                 best_error_list.append(best_val_loss)
                 logged_metric_list.append(train_loss)
                 error_list.append(val_loss)
-                config_list.append({"Current Learner": learner,
-                                    "Current Sample": sample_size,
+                config_list.append({"Current Learner":          learner,
```
Comment (sonichi) [inline flaml/data.py:159]: flask complains these extra spaces. You can find them in "flint with flake8" section of the checks:　https://github.com/microsoft/FLAML/pull/89/checks?check_run_id=2599396437

File: flaml/data.py [hunk 2]
```diff
@@ -197,8 +197,21 @@ def fit_transform(self, X, y, task):
             for column in X.columns:
                 # sklearn\utils\validation.py needs int/float values
                 if X[column].dtype.name == 'datetime64[ns]':
+                    X[f'year_{column}'] = X[column].dt.year
```
Comment (sonichi) [inline flaml/data.py:200]: we need to make sure these newly added columns do not exist in the original dataframe. If they exist, we shouldn't overwrite them.

File: flaml/data.py [hunk 3]
```diff
@@ -253,13 +266,21 @@ def transform(self, X):
...
         if isinstance(X, pd.DataFrame):
             cat_columns, num_columns, datetime_columns = self._cat_columns, \
-                self._num_columns, self._datetime_columns
-            X = X[cat_columns + num_columns].copy()
+                                                         self._num_columns, self._datetime_columns
             if datetime_columns:
-                for dt_column in datetime_columns:
-                    X[dt_column] = X[dt_column].map(datetime.toordinal)
+                for column in datetime_columns:
+                    X[f'year_{column}'] = X[column].dt.year
```
Comment (sonichi) [inline flaml/data.py:272]: assign X[column].dt to a variable for repeated use.

File: flaml/data.py [hunk 4]
```diff
@@ -197,8 +197,21 @@ def fit_transform(self, X, y, task):
             for column in X.columns:
                 # sklearn\utils\validation.py needs int/float values
                 if X[column].dtype.name == 'datetime64[ns]':
```
Comment (sonichi) [inline flaml/data.py:199]: This check should be moved after the check for drop (line 234). Also, the added features should go through the same check for uniqueness before getting added.

File: flaml/data.py (hunk 2 above)
Comment (gianpDomiziani) [reply flaml/data.py:200]: What about using a random prefix, for each column name ?

File: flaml/data.py [hunk 5]
```diff
@@ -219,8 +215,24 @@ def fit_transform(self, X, y, task):
...
+                        if X[column].dtype.name == 'datetime64[ns]':
+                            tmp_dt = X[column].dt
+                            new_columns_dict = {f'year_{column}':      tmp_dt.year, f'month_{column}': tmp_dt.month,
+                                                f'day_{column}':       tmp_dt.day, f'hour_{column}': tmp_dt.hour,
+                                                f'minute_{column}':    tmp_dt.minute, f'second_{column}': tmp_dt.second,
+                                                f'dayofweek_{column}': tmp_dt.dayofweek,
+                                                f'dayofyear_{column}': tmp_dt.dayofyear,
+                                                f'quarter_{column}':   tmp_dt.quarter}
+                            for new_col_name in new_columns_dict.keys():
+                                if new_col_name not in X.columns:
```
Comment (sonichi) [inline flaml/data.py:227]: also need to check the drop condition as in line 199 & 200

File: flaml/data.py [hunk 6]
```diff
@@ -253,13 +265,23 @@ def transform(self, X):
...
+                for column in datetime_columns:
+                    tmp_dt = X[column].dt
+                    new_columns_dict = {f'year_{column}':      tmp_dt.year, f'month_{column}': tmp_dt.month,
+                                        f'day_{column}':       tmp_dt.day, f'hour_{column}': tmp_dt.hour,
+                                        f'minute_{column}':    tmp_dt.minute, f'second_{column}': tmp_dt.second,
+                                        f'dayofweek_{column}': tmp_dt.dayofweek,
+                                        f'dayofyear_{column}': tmp_dt.dayofyear,
+                                        f'quarter_{column}':   tmp_dt.quarter}
+                    for new_col_name in new_columns_dict.keys():
+                        if new_col_name not in X.columns:
```
Comment (sonichi) [inline flaml/data.py:279]: also need to check if the new_col_name is in num_columns in case they are dropped

File: flaml/data.py [hunk 7]
```diff
@@ -219,8 +215,25 @@ def fit_transform(self, X, y, task):
...
+                            tmp_dt = X[column].dt
+                            new_columns_dict = {f'year_{column}':      tmp_dt.year, f'month_{column}': tmp_dt.month,
+                                                f'day_{column}':       tmp_dt.day, f'hour_{column}': tmp_dt.hour,
+                                                f'minute_{column}':    tmp_dt.minute, f'second_{column}': tmp_dt.second,
+                                                f'dayofweek_{column}': tmp_dt.dayofweek,
+                                                f'dayofyear_{column}': tmp_dt.dayofyear,
+                                                f'quarter_{column}':   tmp_dt.quarter}
+                            for new_col_name in new_columns_dict.keys():
+                                if new_col_name not in X.columns and \
+                                        new_columns_dict.get(new_col_name).nunique(dropna=False) > 2:
```
Comment (sonichi) [inline flaml/data.py:228]: >=

File: flaml/data.py [hunk 8]
```diff
@@ -253,13 +266,23 @@ def transform(self, X):
...
+                for column in datetime_columns:
+                    tmp_dt = X[column].dt
+                    new_columns_dict = {f'year_{column}':      tmp_dt.year, f'month_{column}': tmp_dt.month,
+                                        f'day_{column}':       tmp_dt.day, f'hour_{column}': tmp_dt.hour,
+                                        f'minute_{column}':    tmp_dt.minute, f'second_{column}': tmp_dt.second,
+                                        f'dayofweek_{column}': tmp_dt.dayofweek,
+                                        f'dayofyear_{column}': tmp_dt.dayofyear,
+                                        f'quarter_{column}':   tmp_dt.quarter}
+                    for new_col_name in new_columns_dict.keys():
+                        if new_col_name in num_columns:
```
Comment (sonichi) [inline flaml/data.py:280]: need to check if new_col_name is already in X

Comment (gianpDomiziani) [reply flaml/data.py:280]: at this stage, a test matrix is transformed, it means the columns must be equal to the column of the training matrix, so the check i think is redundant, am'i wrong?

**Review by sonichi — CHANGES_REQUESTED** [review]
Please make sure the checks are passed, including the lint check, and all the comments are addressed.

File: flaml/data.py [hunk 9]
```diff
@@ -253,13 +267,24 @@ def transform(self, X):
...
+                                        f'quarter_{column}':   tmp_dt.quarter}
+                    for new_col_name in new_columns_dict.keys():
+                        if new_col_name in num_columns and new_col_name not in X.columns:
+                            X[new_col_name] = new_columns_dict.get(new_col_name)
+                    X[column] = X[column].map(datetime.toordinal)
+                    del tmp_dt
+            X = X[cat_columns + num_columns].copy()
             for column in cat_columns:
-                # print(column, X[column].dtype.name)
+                # print(column, X[column].dtype.name
```
Comment (sonichi) [inline flaml/data.py:287]: remove this line

**Review by sonichi — COMMENTED** [review]
There are a few remaining lint issues and a few comments. I'll approve after they are resolved.

File: flaml/data.py (hunk 7 above)
Comment (sonichi) [reply flaml/data.py:229]: Could you fix this?

File: flaml/data.py (hunk 9 above)
Comment (gianpDomiziani) [reply flaml/data.py]: Sorry, i've run "coverage run -m pytest test" locally and all tests have passed. If you can tell me what types of lint problems must be solved, i can try.

Comment (sonichi) [reply flaml/data.py]: Thanks. This check fails: https://github.com/microsoft/FLAML/pull/89/checks?check_run_id=2639851684

First, make the test pass. Second, fix the lint issues in the lint section:
.\flaml\data.py:270:58: E126 continuation line over-indented for hanging indent
.\flaml\data.py:275:57: E241 multiple spaces after ':'
.\test\test_automl.py:182:9: E122 continuation line missing indentation or outdented
.\test\test_automl.py:236:9: E303 too many blank lines (2)
.\test\test_automl.py:251:5: E303 too many blank lines (2)

Comment (sonichi) [reply flaml/data.py]: @gianpDomiziani I will check in a patch that might fix the test. You can focus on the lint issues now.

Comment (gianpDomiziani) [reply flaml/data.py]: ok, i can't see the error, because locally all tests passed. Why does this happen? anyway, the lint issue should be solved now.

Comment (gianpDomiziani) [reply flaml/data.py]: Ok, the lint issues should be fixed now. Anyway, i can't realize why locally all tests passed.

Comment (sonichi) [reply flaml/data.py]: Thanks. There are still 3 lint issues left:
./test/test_automl.py:248:5: E303 too many blank lines (2)
./flaml/data.py:270:37: E126 continuation line over-indented for hanging indent
./flaml/data.py:275:57: E241 multiple spaces after ':'

Comment (sonichi) [reply flaml/data.py]: one last lint issue:
./flaml/data.py:270:21: E126 continuation line over-indented for hanging indent
I'll approve after that is fixed. Thanks.

Comment (sonichi) [reply flaml/data.py]: also please change version number to 0.4.1

**Review by sonichi — APPROVED** [review]

**Pull request merged**